File: evals/refs/bst.cpp
#include <iostream>
#include <vector>
using namespace std;

struct Node { int v; Node *l, *r; Node(int x) : v(x), l(nullptr), r(nullptr) {} };

Node* insert(Node *root, int v) {
    if (!root) return new Node(v);
    Node *cur = root;
    while (true) {
        if (v < cur->v) {
            if (!cur->l) { cur->l = new Node(v); return root; }
            cur = cur->l;
        } else if (v > cur->v) {
            if (!cur->r) { cur->r = new Node(v); return root; }
            cur = cur->r;
        } else return root;
    }
}

void inorder(Node *root, vector<int> &out) {
    vector<Node*> stack;
    Node *cur = root;
    while (cur || !stack.empty()) {
        while (cur) { stack.push_back(cur); cur = cur->l; }
        cur = stack.back(); stack.pop_back();
        out.push_back(cur->v);
        cur = cur->r;
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    int T; cin >> T;
    while (T--) {
        int N; cin >> N;
        Node *root = nullptr;
        for (int i = 0; i < N; i++) {
            int v; cin >> v;
            root = insert(root, v);
        }
        vector<int> out;
        inorder(root, out);
        for (size_t i = 0; i < out.size(); i++) {
            if (i > 0) cout << ' ';
            cout << out[i];
        }
        cout << '\n';
    }
    return 0;
}

File: evals/refs/bst.zig
const std = @import("std");

const Node = struct {
    v: i64,
    l: ?*Node = null,
    r: ?*Node = null,
};

fn insert(allocator: std.mem.Allocator, root: ?*Node, v: i64) !*Node {
    if (root == null) {
        const n = try allocator.create(Node);
        n.* = .{ .v = v };
        return n;
    }
    var cur = root.?;
    while (true) {
        if (v < cur.v) {
            if (cur.l == null) {
                const n = try allocator.create(Node);
                n.* = .{ .v = v };
                cur.l = n;
                return root.?;
            }
            cur = cur.l.?;
        } else if (v > cur.v) {
            if (cur.r == null) {
                const n = try allocator.create(Node);
                n.* = .{ .v = v };
                cur.r = n;
                return root.?;
            }
            cur = cur.r.?;
        } else return root.?;
    }
}

fn inorder(root: ?*Node, out: *std.ArrayList(i64), allocator: std.mem.Allocator) !void {
    var stack: std.ArrayList(*Node) = .empty;
    defer stack.deinit(allocator);
    var cur = root;
    while (cur != null or stack.items.len > 0) {
        while (cur) |c| {
            try stack.append(allocator, c);
            cur = c.l;
        }
        const c = stack.items[stack.items.len - 1];
        stack.items.len -= 1;
        try out.append(allocator, c.v);
        cur = c.r;
    }
}

pub fn main(init: std.process.Init) !void {
    const arena = init.arena.allocator();
    var in_buf: [4096]u8 = undefined;
    var fr = std.Io.File.stdin().reader(init.io, &in_buf);
    const r = &fr.interface;
    var data: std.ArrayList(u8) = .empty;
    try r.appendRemainingUnlimited(arena, &data);
    var out_buf: [16384]u8 = undefined;
    var fw = std.Io.File.stdout().writer(init.io, &out_buf);
    const w = &fw.interface;
    var it = std.mem.tokenizeAny(u8, data.items, " \n\r\t");
    const t = try std.fmt.parseInt(usize, it.next().?, 10);
    var c: usize = 0;
    while (c < t) : (c += 1) {
        const n = try std.fmt.parseInt(usize, it.next().?, 10);
        var root: ?*Node = null;
        var i: usize = 0;
        while (i < n) : (i += 1) {
            const v = try std.fmt.parseInt(i64, it.next().?, 10);
            root = try insert(arena, root, v);
        }
        var vs: std.ArrayList(i64) = .empty;
        try inorder(root, &vs, arena);
        var k: usize = 0;
        while (k < vs.items.len) : (k += 1) {
            if (k > 0) try w.writeByte(' ');
            try w.print("{d}", .{vs.items[k]});
        }
        try w.writeByte('\n');
    }
    try w.flush();
}
